### DISCUSSION TITLE
Can somebody help me with the refraction bug?

### DISCUSSION THREAD

**Discussion by Bowen951209** [RayTracing/raytracing.github.io#1641] [Q&A] [answered]
Hello, I'm currently working through the Ray Tracing in One Weekend book and implementing it in Java with OpenGL. While implementing the dielectric material, I encountered a strange bug. Here's a screenshot of the issue:
![image](https://github.com/user-attachments/assets/468073df-c420-42c1-89f8-b6e820c2957f)
GLSL has a built-in refract function, so I don't think the issue is there. I suspect the problem might be in the get_color() function, where I attempt to simulate the recursive ray-tracing mechanism as described in the book. However, just in case the problem isn't there, here's a link to my [repository](https://github.com/Bowen951209/raytracing-book)(at commit: 3f28c0e47979a17c4107b327a90e47ce45e928a6)

You can also check out my main compute shader [code](https://github.com/Bowen951209/raytracing-book/blob/refraction/src/main/resources/shaders/raytrace/compute.glsl).

I apologize if this turns out to be a simple issue, but I've been struggling with it for a while and haven't found a solution yet. Any help would be greatly appreciated!

**Comment by whydoubt** [comment, marked as answer]
Note that an eta value of 1.0 should allow a ray to enter the sphere, and exit, without altering the direction of the ray, making the sphere effectively invisible.  If you test this with your code, there are still problems.  So you are correct that the refract is not the problem.

I found two places that seem to be making this not behave properly:
- In `near_zero()` you need to checking against `abs()` of each component.  That is causing the bit of blue sphere visible in the lower half but not the upper.
- In `get_color()`, you only continue/recurse if `dot(ray.dir, hit_record.normal) > 0.0`.  Inside the `if (has_hit_anything)` block, I believe you should _always_ continue/recurse.

**Comment by Bowen951209** [comment]
Thank you! The issue was indeed in the `near_zero()` function and the section of the `get_color()` function that you pointed out. By the way, the `dot(ray.dir, hit_record.normal) > 0.0` check should be only for metal materials.
